File: leetcode/c#/Problems/P2416.cs
namespace LeetCode.Naive.Problems;

/// <summary>
///    Problem: https://leetcode.com/problems/sum-of-prefix-scores-of-strings/
///    Submission: https://leetcode.com/submissions/detail/803573522/
/// </summary>
internal class P2416
{
  public class Solution
  {
    public int[] SumPrefixScores(string[] words)
    {
      var trie = new Trie();

      foreach (var word in words)
      {
        trie.Add(word);
      }

      var ans = new int[words.Length];

      for (var i = 0; i < words.Length; i++)
      {
        var word = words[i];
        var count = 0;

        var node = trie.Root;
        foreach (var ch in word)
        {
          count += node.Nodes[ch].Score;
          node = node.Nodes[ch];
        }

        ans[i] = count;
      }

      return ans;
    }

    public class Trie
    {
      public class TrieNode
      {
        public int Value;
        public int Score;
        public Dictionary<char, TrieNode> Nodes { get; } = new Dictionary<char, TrieNode>();

        public TrieNode(int value)
        {
          Value = value;
        }

        public TrieNode Add(char value)
        {
          if (Nodes.ContainsKey(value))
          {
            Nodes[value].Score++;
            return Nodes[value];
          }

          var node = new TrieNode(value);
          node.Score++;
          Nodes[value] = node;

          return node;
        }
      }

      public TrieNode Root { get; } = new TrieNode(default);

      public void Add(IEnumerable<char> seq)
      {
        var current = Root;

        foreach (var element in seq)
        {
          current = current.Add(element);
        }
      }
    }
  }
}

File: leetcode/python/P2416.py
"""
Problem: https://leetcode.com/problems/sum-of-prefix-scores-of-strings/
Solution: https://leetcode.com/submissions/detail/803580405/
"""

from typing import List

class Solution:
  def sumPrefixScores(self, words: List[str]) -> List[int]:
    trie = Solution.Trie()

    for w in words:
      trie.add(w)

    ans = []

    for w in words:
      count = 0
      node = trie.root
      for ch in w:
        count += node.nodes[ch].score
        node = node.nodes[ch]
      ans.append(count)

    return ans

  class TrieNode:
    def __init__(self, ch, score):
      self.ch = ch
      self.score = score
      self.nodes = dict()

    def add(self, ch):
      if self.nodes.get(ch):
        self.nodes[ch].score += 1
        return self.nodes[ch]

      node = Solution.TrieNode(ch, 1)
      self.nodes[ch] = node
      return node

  class Trie:
    def __init__(self):
      self.root = Solution.TrieNode(None, 0)

    def add(self, s: str):
      current = self.root
      for ch in s:
        current = current.add(ch)